File: mac/Voice/Refine/Outbox.swift
import Foundation

struct OutboxEntry {
    // G2: optional — literal-mode dictations carry no fallback reason at all (nil, not `.offline`).
    let clientId: UUID; let raw: String; let injected: Injected; let fallback: FallbackReason?; let createdAt: Date
    let mode: String; let languageSetting: String; let languageDetected: String?; let app: FrontmostApp?; let audioMs: Int; let asrMs: Int
    // prd-sub-second-dictation.md rule 18: an offline dictation still has to contribute its
    // measurement once connectivity returns, so the timing rides the queued entry rather than
    // being recomputed at replay time — by then the original release is long gone.
    var cleaned: String? = nil; var llmMs: Int? = nil; var llmModel: String? = nil; var totalMs: Int? = nil
}

/// Memory only (no transcript ever touches disk). Capped; oldest dropped first.
/// G5: main-actor — constructed and used only from the @MainActor Coordinator/RefineService.
@MainActor
final class Outbox {
    static let cap = 200
    private var entries: [OutboxEntry] = []
    var count: Int { entries.count }
    func add(_ e: OutboxEntry) { entries.append(e); if entries.count > Self.cap { entries.removeFirst(entries.count - Self.cap) } }
    func drain() -> [OutboxEntry] { defer { entries.removeAll() }; return entries }
    func requeue(_ es: [OutboxEntry]) { for e in es { add(e) } }
}

File: windows/Spit.Core/Refine/Outbox.cs
namespace Spit.Core;

/// An unsent dictation report. Port of `OutboxEntry` in mac/Voice/Refine/Outbox.swift.
/// `Fallback` is nullable: literal-mode dictations carry no fallback reason at all (Mac G2).
public sealed record OutboxEntry(
    Guid ClientId,
    string Raw,
    Injected Injected,
    FallbackReason? Fallback,
    DateTimeOffset CreatedAt,
    string Mode,
    string LanguageSetting,
    string? LanguageDetected,
    FrontmostApp? App,
    int AudioMs,
    int AsrMs)
{
    // An offline dictation still has to contribute its measurement once connectivity returns, so the
    // timing rides the queued entry rather than being recomputed at replay time.
    public string? Cleaned { get; init; }
    public int? LlmMs { get; init; }
    public string? LlmModel { get; init; }
    public int? TotalMs { get; init; }
}

/// Memory only — no transcript ever touches disk (rule 25). Capped; oldest dropped first.
/// Port of mac/Voice/Refine/Outbox.swift; locked because Windows has no main actor to serialise it.
public sealed class Outbox
{
    public const int Cap = 200;

    private readonly List<OutboxEntry> _entries = [];
    private readonly Lock _lock = new();

    public int Count
    {
        get { lock (_lock) return _entries.Count; }
    }

    public void Add(OutboxEntry entry)
    {
        lock (_lock)
        {
            _entries.Add(entry);
            if (_entries.Count > Cap) _entries.RemoveRange(0, _entries.Count - Cap);
        }
    }

    public IReadOnlyList<OutboxEntry> Drain()
    {
        lock (_lock)
        {
            var all = _entries.ToArray();
            _entries.Clear();
            return all;
        }
    }

    public void Requeue(IEnumerable<OutboxEntry> entries)
    {
        foreach (var e in entries) Add(e);
    }
}
